src/algebraic_geometry/EllipticCurve.lean

```diff
@@ -106,38 +106,88 @@ lemma c_relation : 1728 * ↑E.Δ = E.c₄ ^ 3 - E.c₆ ^ 2 := by { simp, ring1
 
 end quantity
 
-/-! ### `2`-torsion polynomials -/
-
 section torsion_polynomial
 
-variables (A : Type v) [comm_ring A] [algebra R A]
+/-! ### `2`-torsion polynomials -/
 
 /-- The polynomial whose roots over a splitting field of `R` are the `2`-torsion points of the
   elliptic curve when `R` is a field of characteristic different from `2`, and whose discriminant
   happens to be a multiple of the discriminant of the elliptic curve. -/
-def two_torsion_polynomial : cubic A :=
-⟨4, algebra_map R A E.b₂, 2 * algebra_map R A E.b₄, algebra_map R A E.b₆⟩
+def two_torsion_polynomial : cubic R := ⟨4, E.b₂, 2 * E.b₄, E.b₆⟩
 
-lemma two_torsion_polynomial.disc_eq :
-  (two_torsion_polynomial E A).disc = 16 * algebra_map R A E.Δ :=
-begin
-  simp only [two_torsion_polynomial, cubic.disc, coe_Δ, b₂, b₄, b₆, b₈, map_neg, map_add, map_sub,
-             map_mul, map_pow, map_one, map_bit0, map_bit1],
-  ring1
-end
+lemma two_torsion_polynomial.disc_eq : E.two_torsion_polynomial.disc = 16 * E.Δ :=
+by { simp only [two_torsion_polynomial, cubic.disc, coe_Δ, b₂, b₄, b₆, b₈], ring1 }
 
 lemma two_torsion_polynomial.disc_ne_zero {K : Type u} [field K] [invertible (2 : K)]
-  (E : EllipticCurve K) (A : Type v) [comm_ring A] [nontrivial A] [algebra K A] :
-  (two_torsion_polynomial E A).disc ≠ 0 :=
+  (E : EllipticCurve K) : E.two_torsion_polynomial.disc ≠ 0 :=
 λ hdisc, E.Δ.ne_zero $ mul_left_cancel₀ (pow_ne_zero 4 $ nonzero_of_invertible (2 : K)) $
-  (algebra_map K A).injective
-begin
-  simp only [map_mul, map_pow, map_bit0, map_one, map_zero],
-  linear_combination hdisc - two_torsion_polynomial.disc_eq E A
-end
+by linear_combination hdisc - two_torsion_polynomial.disc_eq E
+    with { normalization_tactic := `[ring1] }
 
 end torsion_polynomial
 
+section base_change
+
+/-! ### Base change -/
+
+variables (A : Type v) [comm_ring A] [algebra R A]
+
+private meta def simp_map : tactic unit :=
+`[simp only [map_one, map_bit0, map_bit1, map_neg, map_add, map_sub, map_mul, map_pow]]
+
+/-- The elliptic curve over `R` base changed to `A`. -/
+def base_change : EllipticCurve A :=
+{ a₁   := algebra_map R A E.a₁,
+  a₂   := algebra_map R A E.a₂,
+  a₃   := algebra_map R A E.a₃,
+  a₄   := algebra_map R A E.a₄,
+  a₆   := algebra_map R A E.a₆,
+  Δ    := units.map ↑(algebra_map R A) E.Δ,
+  Δ_eq := by { simp only [units.coe_map, ring_hom.coe_monoid_hom, Δ_eq, Δ_aux], simp_map } }
+
+namespace base_change
+
+@[simp] lemma a₁_eq : (E.base_change A).a₁ = algebra_map R A E.a₁ := rfl
+
+@[simp] lemma a₂_eq : (E.base_change A).a₂ = algebra_map R A E.a₂ := rfl
+
+@[simp] lemma a₃_eq : (E.base_change A).a₃ = algebra_map R A E.a₃ := rfl
+
+@[simp] lemma a₄_eq : (E.base_change A).a₄ = algebra_map R A E.a₄ := rfl
+
+@[simp] lemma a₆_eq : (E.base_change A).a₆ = algebra_map R A E.a₆ := rfl
```

Comment:
I think putting `@[simps]` on `def base_change` gives you these for free?

Reply:
Yeah, thanks. I thought the `namespace` allows for dot notation but it doesn't work and it's probably not useful anyway. Here is the PR that addresses both this and `change_of_variable`: #17708